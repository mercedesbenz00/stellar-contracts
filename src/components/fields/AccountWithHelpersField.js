import React from 'react'
import {Button, Col, FormControl, Row} from 'react-bootstrap'
import PropTypes from 'prop-types'
import sdk from 'stellar-sdk'

/**
 * Stellar account field (for jsonschema forms) that adds:
 *  - Generate button - to create a new account
 *  - Use Signer button - to use account signed into the app
 */
class AccountWithHelpersField extends React.Component {
  constructor(props) {
    super(props)
    this.state = {...props.formData}
  }

  handleOnClickGenerate = () => {
    const newKeypair = sdk.Keypair.random()
    this.setState(
      {
        secretKey: newKeypair.secret(),
      },
      () => this.props.onChange(this.state)
    )
  }

  handleOnClickUseSigner = signer => {
    this.setState({secretKey: signer}, () => this.props.onChange(this.state))
  }

  handleOnChange = e => {
    console.log(`AccountWithHelpersField: onChange: ${e.target.value}`)
    this.setState({secretKey: e.target.value}, () =>
      this.props.onChange(this.state)
    )
  }

  render() {
    return (
      <Row>
        <Col xs={8}>
          <FormControl
            onChange={this.handleOnChange}
            placeholder={this.props.uiSchema['ui:placeholder']}
            type="text"
            value={this.state.secretKey}
          />
        </Col>
        <Col xs={2}>
          <Button
            id="btn-generate"
            bsStyle="success"
            onClick={this.handleOnClickGenerate}
          >
            Generate
          </Button>
        </Col>
        <Col xs={2}>
          <Button
            id="btn-use-signer"
            bsStyle="success"
            onClick={() =>
              this.handleOnClickUseSigner(this.props.formContext.signer)}
          >
            Use Signer
          </Button>
        </Col>
      </Row>
    )
  }
}

AccountWithHelpersField.propTypes = {
  formContext: PropTypes.shape({
    signer: PropTypes.string,
  }).isRequired,
  formData: PropTypes.object,
  onChange: PropTypes.func.isRequired,
}

export default AccountWithHelpersField
