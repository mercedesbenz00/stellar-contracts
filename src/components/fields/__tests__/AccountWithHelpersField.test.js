import React from 'react'
import {shallow} from 'enzyme'
import {Keypair, StrKey} from 'stellar-sdk'

import AccountWithHelpersField from '../AccountWithHelpersField'

it('renders with minimal fields', () => {
  const onChange = jest.fn()
  const field = shallow(
    <AccountWithHelpersField
      formContext={{}}
      formData={{}}
      onChange={onChange}
      uiSchema={{}}
    />
  )
  expect(field.getNodes()).toMatchSnapshot()
})

it('renders placeholder', () => {
  const onChange = jest.fn()
  const placeholder = 'Enter a Stellar Secret Key'
  const field = shallow(
    <AccountWithHelpersField
      formContext={{}}
      formData={{}}
      onChange={onChange}
      uiSchema={{'ui:placeholder': placeholder}}
    />
  )
  expect(field.find('[type="text"]').prop('placeholder')).toEqual(placeholder)
  expect(field.getNodes()).toMatchSnapshot()
})

it('generate button inserts a new account', () => {
  const onChange = jest.fn()
  const field = shallow(
    <AccountWithHelpersField
      formContext={{}}
      formData={{}}
      onChange={onChange}
      uiSchema={{}}
    />
  )
  field.find('#btn-generate').simulate('click')

  const secretKey = field.find('[type="text"]').prop('value')
  expect(StrKey.isValidEd25519SecretSeed(secretKey)).toEqual(true)
  expect(onChange).toHaveBeenCalledWith({secretKey: secretKey})
})

it('use signer button inserts the signer', () => {
  const keypair = Keypair.random()
  const formCtx = {signer: keypair.secret()}
  const onChange = jest.fn()
  const field = shallow(
    <AccountWithHelpersField
      formContext={formCtx}
      formData={{}}
      onChange={onChange}
      uiSchema={{}}
    />
  )
  field.find('#btn-use-signer').simulate('click')

  const secretKey = field.find('[type="text"]').prop('value')
  expect(secretKey).toEqual(keypair.secret())
  expect(onChange).toHaveBeenCalledWith({secretKey: secretKey})
})
