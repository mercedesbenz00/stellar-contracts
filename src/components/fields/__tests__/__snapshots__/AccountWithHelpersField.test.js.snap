// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders placeholder 1`] = `
Array [
  <Row
    bsClass="row"
    componentClass="div"
>
    <Col
        bsClass="col"
        componentClass="div"
        xs={8}
    >
        <FormControl
            bsClass="form-control"
            componentClass="input"
            onChange={[Function]}
            placeholder="Enter a Stellar Secret Key"
            type="text"
            value={undefined}
        />
    </Col>
    <Col
        bsClass="col"
        componentClass="div"
        xs={2}
    >
        <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="success"
            disabled={false}
            id="btn-generate"
            onClick={[Function]}
        >
            Generate
        </Button>
    </Col>
    <Col
        bsClass="col"
        componentClass="div"
        xs={2}
    >
        <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="success"
            disabled={false}
            id="btn-use-signer"
            onClick={[Function]}
        >
            Use Signer
        </Button>
    </Col>
</Row>,
]
`;

exports[`renders with minimal fields 1`] = `
Array [
  <Row
    bsClass="row"
    componentClass="div"
>
    <Col
        bsClass="col"
        componentClass="div"
        xs={8}
    >
        <FormControl
            bsClass="form-control"
            componentClass="input"
            onChange={[Function]}
            placeholder={undefined}
            type="text"
            value={undefined}
        />
    </Col>
    <Col
        bsClass="col"
        componentClass="div"
        xs={2}
    >
        <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="success"
            disabled={false}
            id="btn-generate"
            onClick={[Function]}
        >
            Generate
        </Button>
    </Col>
    <Col
        bsClass="col"
        componentClass="div"
        xs={2}
    >
        <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="success"
            disabled={false}
            id="btn-use-signer"
            onClick={[Function]}
        >
            Use Signer
        </Button>
    </Col>
</Row>,
]
`;
